src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -992,7 +1176,23 @@ void ShenandoahConcurrentGC::op_final_updaterefs() {
 }
 
 void ShenandoahConcurrentGC::op_final_roots() {
-  ShenandoahHeap::heap()->set_concurrent_weak_root_in_progress(false);
+
+  ShenandoahHeap *heap = ShenandoahHeap::heap();
+  heap->set_concurrent_weak_root_in_progress(false);
+  heap->set_evacuation_in_progress(false);
```

Comment:
This will take a little explaining...

The generational mode allows entire regions to be "promoted in place". This happens during the evacuation phase -_even if there is nothing to evacuate_. So, in the case when the collection set is empty, but we have regions we want to promote in place we turn this flag on to progress through the evacuation phase. However, in this case, we will still skip the update references phase which ordinarily clears the evacuation in progress phase. So, to preserve the expected value of this flag for these short-cut cycles that promote in place, we clear the evacuation in progress flag here in `op_final_roots`.

Reply:
ok, thank you!